File: 0x08-python-more_classes/4-rectangle.py
#!/usr/bin/python3
"""a class Rectangle that defines a rectangle"""


class Rectangle:
    """Instantiation with optional width and height"""

    def __init__(self, width=0, height=0):
        """defining width and height"""

        self.width = width
        self.height = height

    @property
    def width(self):
        """using getter to retrieve the width"""

        return self.__width

    @width.setter
    def width(self, value):
        """using setter to set exceptions"""
        if type(value) is not int:
            raise TypeError("width must be an integer")
        if value < 0:
            raise ValueError("width must be >= 0")
        self.__width = value

    @property
    def height(self):
        """using getter to retrieve the height"""

        return self.__height

    @height.setter
    def height(self, value):
        """using setter to set exceptions"""

        if type(value) is not int:
            raise TypeError("height must be an integer")
        if value < 0:
            raise ValueError("height must be >= 0")
        self.__height = value

    def area(self):
        """instance method that returns the rectangle area"""

        return self.__width * self.__height

    def perimeter(self):
        """instance method that returns the rectangle perimeter"""

        if self.__width == 0 or self.__height == 0:
            return 0
        else:
            return (self.__width * 2) + (self.__height * 2)

    def __str__(self):
        """print the rectangle with the character #"""

        s = ""
        if self.__width != 0 and self.__height != 0:
            s = s + "\n".join("#" * self.__width for h in range(self.__height))
        return s

    def __repr__(self):
        """return a string representation of the rectangle"""

        return "Rectangle({:d}, {:d})".format(self.__width, self.__height)

File: 0x13-javascript_objects_scopes_closures/4-rectangle.js
#!/usr/bin/node
// a class Rectangle that defines a rectangle
// must use the class notation for defining your class
// The constructor must take 2 arguments w and h
// Initialize the instance attribute width with the value of w
// Initialize the instance attribute height with the value of h
// If w or h is equal to 0 or not a positive integer, create an empty object
// Create an instance method called print() that prints the rectangle using the character X
// Create an instance method called rotate() that exchanges the width and the height of the rectangle
// Create an instance method called double() that multiples the width and the height of the rectangle by 2
class Rectangle {
  constructor (w, h) {
    if ((w > 0) && (h > 0)) {
      this.width = w;
      this.height = h;
    }
  }

  print () {
    let n = 0;
    while (n < this.height) {
      let r = '';
      n++;
      for (let m = 0; m < this.width; m++) {
        r = r + 'X';
      }
      console.log(r);
    }
  }

  rotate () {
    const exg = this.width;
    this.width = this.height;
    this.height = exg;
  }

  double () {
    this.width = this.width * 2;
    this.height = this.height * 2;
  }
}
module.exports = Rectangle;
